File: jetbrains-plugin/src/main/kotlin/com/ezevillo/gitreview/domain/StaleGuard.kt
package com.ezevillo.gitreview.domain

data class StateToken(
    val branch: String? = null,
    val tip: String? = null,
    val situation: Situation,
)

fun captureToken(state: ReviewState): StateToken = StateToken(
    branch = state.state?.branch,
    tip = state.state?.tip,
    situation = state.situation,
)

fun tokenStillValid(token: StateToken, state: ReviewState): Boolean =
    token.situation == state.situation &&
        token.branch == state.state?.branch &&
        token.tip == state.state?.tip

File: visualstudio-extension/src/GitReview.Domain/StaleGuard.cs
namespace GitReview.Domain;

public sealed record StateToken(
    string? Branch,
    string? Tip,
    Situation Situation);

public static class StaleGuard
{
    public static StateToken CaptureToken(ReviewState state) => new(
        state.State?.Branch,
        state.State?.Tip,
        state.Situation);

    public static bool TokenStillValid(StateToken token, ReviewState state) =>
        token.Situation == state.Situation
        && token.Branch == state.State?.Branch
        && token.Tip == state.State?.Tip;
}
